include/ygm/container/detail/bag_impl.hpp

```diff
@@ -47,6 +82,35 @@ class bag_impl {
     return m_comm.all_reduce_sum(m_local_bag.size());
   }
 
+  size_t local_size() {
+    return m_local_bag.size();
+  }
+
+
+  void rebalance() {
+    m_comm.barrier();
+
+    // Find current rank's prefix val and desired target size
+    size_t prefix_val = ygm::prefix_sum(local_size(), m_comm);
+    size_t target_size = std::ceil((size() * 1.0) / m_comm.size());
+
+    // Init to_send array where index is dest and value is the num to send
+    int to_send[m_comm.size()] = {0};
```

Comment:
Instead of an array, use an unordered_map. This avoids allocating O(p) data on every rank. Each rank will probably be sending to a small number of other ranks, so this will be really small.